Modules/_copy.c

```diff
@@ -0,0 +1,450 @@
+#ifndef Py_BUILD_CORE_BUILTIN
+#  define Py_BUILD_CORE_MODULE 1
+#endif
+
+#include "Python.h"
+#include "pycore_moduleobject.h"  // _PyModule_GetState()
+#include "clinic/_copy.c.h"
+
+/*[clinic input]
+module _copy
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=b34c1b75f49dbfff]*/
+
+#define object_id PyLong_FromVoidPtr
+
+typedef struct {
+    PyObject *python_copy_module;
+    PyObject *str_deepcopy_fallback;
+} copy_module_state;
+
+static inline copy_module_state*
+get_copy_module_state(PyObject *module)
+{
+    void *state = _PyModule_GetState(module);
+    assert(state != NULL);
+    return (copy_module_state *)state;
+}
+
+static int
+memo_keepalive(PyObject* x, PyObject* memo)
+{
+    PyObject *memoid = object_id(memo);
+    if (memoid == NULL)
+        return -1;
+
+    /* try: memo[id(memo)].append(x) */
+    PyObject *list = PyDict_GetItem(memo, memoid);
+    if (list != NULL) {
+        Py_DECREF(memoid);
+        return PyList_Append(list, x);
+    }
+
+    /* except KeyError: memo[id(memo)] = [x] */
+    list = PyList_New(1);
+    if (list == NULL) {
+        Py_DECREF(memoid);
+        return -1;
+    }
+    Py_INCREF(x);
+    PyList_SET_ITEM(list, 0, x);
+    int ret = PyDict_SetItem(memo, memoid, list);
+    Py_DECREF(memoid);
+    Py_DECREF(list);
+    return ret;
+}
+
+/* Forward declaration. */
+static PyObject* do_deepcopy(PyObject *module, PyObject* x, PyObject* memo);
+
+static PyObject*
+do_deepcopy_fallback(PyObject* module, PyObject* x, PyObject* memo)
+{
+    copy_module_state *state = get_copy_module_state(module);
+    PyObject *copymodule = state->python_copy_module;
+    assert(copymodule != NULL);
+
+    return PyObject_CallMethodObjArgs(copymodule, state->str_deepcopy_fallback, x, memo, NULL);
```

Comment:
You can use vectorcall to avoid variadic function overhead.

Reply:
`PyObject_VectorcallMethod` can be used here or you can directly cache the function object and call it with `PyObject_Vectorcall`.

Reply:
I changed the implementation to PyObject_VectorcallMethod